experiments/yarn/src/main/scala/org/apache/gearpump/experiments/yarn/master/YarnApplicationMaster.scala

```diff
@@ -0,0 +1,338 @@
+  }
+
+  private def needMoreWorkerContainersState: Boolean = {
+    workerContainersStarted < appConfig.getEnv(WORKER_CONTAINERS).toInt
+  }
+
+  private def requestWorkerContainersIfNeeded(): Unit = {
+    if (masterContainersStarted == appConfig.getEnv(GEARPUMPMASTER_CONTAINERS).toInt) {
+      LOG.info("Requesting worker containers")
+      requestWorkerContainers()
+    }
+  }
+
+  private def launchMasterContainers(containers: List[Container]) {
+    LOG.info(s"Containers size ${containers.size}")
+    containers.foreach(container => {
+      val port = appConfig.getEnv(GEARPUMPMASTER_PORT).toInt
+      val masterCommand = MasterContainerCommand(appConfig, masterAddr.get)
+      launchCommand(container, masterCommand.getCommand)
+      masterContainers += container.getId ->(container.getNodeId.getHost, port)
+    })
+  }
+
+  private def launchWorkerContainers(containers: List[Container]) {
+    containers.foreach(container => {
+      val workerHost = container.getNodeId.getHost
+      val workerCommand = WorkerContainerCommand(appConfig, masterAddr.get, workerHost).getCommand
+      launchCommand(container, workerCommand)
+    })
+  }
+
+  private def launchCommand(container: Container, command: String) {
+    LOG.info(s"Launching containter: containerId :  ${container.getId}, host ip : ${container.getNodeId.getHost}")
+    LOG.info("Launching command : " + command)
+    val containerContext = containerLaunchContextFactory(command)
+    LOG.info(s"NodeManagerClient : $nodeManagerClient, container : $container, containerContext $containerContext")
+    nodeManagerClient.startContainerAsync(container, containerContext)
+  }
+
+  private def requestWorkerContainers(): Unit = {
+    (1 to appConfig.getEnv(WORKER_CONTAINERS).toInt).foreach(requestId => {
+      rmClientActor ! ContainerRequestMessage(appConfig.getEnv(WORKER_MEMORY).toInt, appConfig.getEnv(WORKER_VCORES).toInt)
+    })
+  }
+
+  private def requestMasterContainers(registrationResponse: RegisterApplicationMasterResponse) = {
+    //TODO remove that or do something more with that
+    val previousContainersCount = registrationResponse.getContainersFromPreviousAttempts.size
+
+    LOG.info(s"Previous container count : $previousContainersCount")
+    if (previousContainersCount > 0) {
+      LOG.warn("Previous container count > 0, can't do anything with it")
+    }
+
+    LOG.info(s"GEARPUMPMASTER_CONTAINERS: ${appConfig.getEnv(GEARPUMPMASTER_CONTAINERS).toInt}")
+    (1 to appConfig.getEnv(GEARPUMPMASTER_CONTAINERS).toInt).foreach(requestId => {
+      rmClientActor ! ContainerRequestMessage(appConfig.getEnv(GEARPUMPMASTER_MEMORY).toInt, appConfig.getEnv(GEARPUMPMASTER_VCORES).toInt)
+    })
+
+  }
+
+  private def cleanUp(done: RMHandlerDone): Boolean = {
+    LOG.info("Application completed. Stopping running containers")
+    nodeManagerClient.stop()
+    var success = true
+
+    val stats = done.rMHandlerContainerStats
+    done.reason match {
```

Comment:
Again, I dislike the "done.reason" case match here.  Can we use class hiearachy instead of member fields?

For example: 
sealed trait ResourceManagerHandlerStatus
class ResourceManagerHandlerFailed extends ResourceManagerHandlerStatus

By the way, I cannot understand what RmHandler actually mean.